ebpf_updates: issue #2 (#91)
cilium/ebpf.io: src/posts/2020-12-21-ebpf-updates/index.md

  * The Containers devroom
    ([CFP](https://discuss.linuxcontainers.org/t/fosdem-2021-containers-devroom-call-for-papers/9625))
    accepts submissions until the 22th of December 2020.

Recent start-up acquisitions highlight the growing adoption and the maturity of
eBPF:
* [Flowmill](https://www.flowmill.com/), offering a solution for network
  observability relying on eBPF,
  [has been acquired](https://techcrunch.com/2020/11/24/splunk-acquires-network-observability-service-flowmill/)
  by [Splunk](https://www.splunk.com/en_us/newsroom/press-releases/2020/splunk-to-acquire-network-performance-monitoring-leader-flowmill.html).
* [Pixie Labs](https://pixielabs.ai/), which uses eBPF for visibility in
  Kubernetes,
  [has been acquired](https://techcrunch.com/2020/12/10/new-relic-acquires-kubernetes-observability-platform-pixie-labs/)
  by [New Relic](https://blog.newrelic.com/product-news/pixie-developer-first-observability/).

Readers from Brazil may be interested in the [eBPF Brasil](https://ebpfbr.org/)
website, which aims at gathering, translating, and sharing knowledge about
eBPF.

## New Resources

### Blog Posts, Presentations

> * [_The Top Reasons Why You Should Give eBPF a Chance_](https://blog.container-solutions.com/the-top-reasons-why-you-should-give-ebpf-a-chance)

Author: Definitely an omission, thank you!

  function in the kernel with little impact on performance. Another reason is
  that more and more companies, including large ones, are adopting eBPF; a
  detailed list follows. At last, eBPF makes it possible to quickly develop
  tools to instrument new parts of the kernel, without the need to go through a
  longer process to upstream new attach points.

* [_Making BPF easy with libbpf and Zig_](https://cmd.com/blog/making-bpf-easy-with-libbpf-and-zig/),
  from Matt Knight.  
  The [Zig](https://ziglang.org) programming language is used to create and
  handle a simple eBPF program in this tutorial, both for the user space loader
  and the eBPF program itself. The objective is mostly to understand how libbpf
  manipulates the ELF object containing a program in order to load it, but
  compiling from Zig, which aims at competing with C while offering some newer
  features, may open new perspectives. The code is available
  [on GitHub](https://github.com/mattnite/zig-bpf-intro).

* [_Beyond the Buzzword: BPF’s Unexpected Role in Kubernetes_](https://kccncna20.sched.com/event/ekDR/beyond-the-buzzword-bpfs-unexpected-role-in-kubernetes-andrew-randall-alban-crequy-kinvolk),

>   from Andrew Randall Alban Crequy.  

Reviewer: ```suggestion
  from Andrew Randall and Alban Crequy.  
```

  After a high-level overview of eBPF, this presentation depicts the landscape
  of the projects gravitating around this technology. The authors explain that
  there are many powerful tools based on eBPF, although none of them would
  cover Kubernetes clusters. As an answer to fill the gap, they introduce
  Inspektor Gadget, which reuses some elements from the bcc tools to provide a

>   new set of monitoring gadgets for examining Pods. Note that the wording might

Reviewer 2: ```suggestion
  new set of monitoring gadgets for examining Pods. Note that the wording
```

  in the slides might be misleading: If there was no equivalent to bcc for

>   tracing containers before Inspecktor Gadget, there _are_ other tools

Reviewer 2: ```suggestion
  tracing containers before Inspektor Gadget, there _are_ other tools
```

  from Juraj Selep.  
  TezEdge peer-to-peer nodes validate blocks for the decentralized
  [Tezos](https://en.wikipedia.org/wiki/Tezos) blockchain, providing smart
  contracts. Blockchain networks are subject to DDoS (Distributed Denial of
  Service) attacks, generally mitigated with a firewall. In the current case,
  XDP is used to implement it. The eBPF program checks that each connection
  starts with a valid and unique proof of work, making it expensive for an
  adversary to start many connections. This is further integrated with the
  “multipass validation” scheme that aims at detecting erroneous blocks as soon
  as possible. Note that the eBPF programs are written in Rust.

* [_Snooping MySQL passwords with eBPF_](https://distributeddba.com/blog/database-administration/data-security/snooping-mysql-passwords-with-ebpf/),
  from Derek Downey.  
  Providing unencrypted password through command prompt or environment
  variables may present a security issue. This short post relies on bcc's
  execsnoop tool to show that MySQL passwords can be retrieved by examining the

>   arguments passed to the system call, on the kernel side.

Reviewer 3: Not sure we should include that one. It's demonstrating why passing passwords in clear text to MySQL is insecure with bcc's execsnoop. I don't think that makes sense given execsnoop requires root privileges. Using the interaction prompt to provide the password to MySQL is not going to protect against a privileged user either.

More generally, I'm afraid it could give the wrong impression that BPF tracing tools are dangerous and can be used for privilege escalation.

Author: I was a bit hesitant as well. Ok I'll remove it.

  from Wenbo Zhang.  
  Another article on the benefits brought by CO-RE, for which libbpf provides
  good support. After comparing bcc-based and libbpf-based tracing tools in
  terms of features and memory footprint, the author provide a list of tools
  and invocation patterns they use to analyze I/O performance.

* [_Tips and Tricks for Writing Linux BPF Applications with libbpf_](https://pingcap.com/blog/tips-and-tricks-for-writing-linux-bpf-applications-with-libbpf),
  from Wenbo Zhang.  
  This is a collection of hints about libbpf, to make it easier to write
  applications or to get better performance. The article contains several hints
  related to the use of eBPF “skeletons” generated by bpftool, eBPF maps, eBPF
  global variables, or pointers dereferencing.

* [_BPF and Go: Modern forms of introspection in Linux_](https://medium.com/bumble-tech/bpf-and-go-modern-forms-of-introspection-in-linux-6b9802682223),
  from Marko Kevac.  
  This long read contains two parts. The first one is a good introduction, in
  terms of context and history, to eBPF. The second part focuses on the

>   interactions between eBPF. If it is not possible to write eBPF programs in Go

Reviewer 3: `interactions between eBPF` and what?

Author: :grin:

  at this date, several Go libraries are available to write user applications
  to manage eBPF objects in this language. As for observability, Go has some

>   particularities that make some it harder or impossible to get thread id for

Reviewer 3: `that make it somewhat harder`?

  goroutines or use uretprobes. But the article demonstrates that a large part
  of eBPF's tracing capabilities remain available for Go applications.

* [_Adding BPF target support to the Rust compiler_](https://confused.ai/posts/rust-bpf-target),
  from Alessandro Decina.  

>   eBPF bytecode can be compiled from C with clang/LLVM's backend. Other

Reviewer 3: ```suggestion
  eBPF bytecode can be compiled from C with Clang/LLVM's backend. Other
```

  languages, and Rust in particular, use the same toolchain. Compiling Rust

>   code into eBPF is not straightforward, but some hacks permit it. This post

Author: I'll rephrase, but “it can be done” sounds very much like a passive form, Paul would never forgive me! :)

  provides details on the linking steps, where the issue lies, and introduces a
  new linker and two related compiler targets for rustc, to ease compilation
  from Rust to eBPF bytecode. Although the targets are not merged yet, a
  command-line hack allows developers to use it with the stable versions of
  rustc and cargo. The linker is available [on
  GitHub](https://github.com/alessandrod/bpf-linker).

### Academic Research

> * [_Express Data Path Kernel Objects for Real-Time Audio Streaming Optimization_](https://lac2020.sciencesconf.org/data/proceedings.pdf#section*.12),

Reviewer 3: This is quite a big shootout for a paper to a non-international conference. Could we at least clarify the conference in the first line (`Linux Audio Conference 2020`)?

Author: Can do, or do you think I should simply drop the subsection and move it above with the other presentations maybe?

Reviewer 3: Yeah, that would also be a solution IMO. I tend to hold academic papers to higher/different standards than blog posts, and in particular, I wouldn't want us to put e.g. a good OSDI paper and a non-international workshop paper on the same level.

  busy-polling to be performed on XDP sockets.
  (Björn Töpel,
  [link](https://lore.kernel.org/bpf/20201130185205.196029-1-bjorn.topel@gmail.com/t/#u))

* Add a new helper `bpf_ima_inode_hash()` to get the IMA (Integrity Measurement
  Architecture) hash of an inode, which can be useful when using eBPF in a LSM
  (Linux Security Module) for fingerprinting files. One example use case is to
  get and store fingerprints of executable files when they are executed, and to
  detect when they attempt to unlink themselves.
  (KP Singh,
  [link](https://lore.kernel.org/bpf/20201124151210.1081188-1-kpsingh@chromium.org/t/#u))

* Add a new helper `bpf_bprm_opts_set()` for use with eBPF in a LSM, to update
  certain bits of the `struct linux_binprm` for a process. The only accessible
  bit at the moment is the `secureexec` bit, which can (indirectly) disable the
  use of certain environment variables like `LD_PRELOAD` for the dynamic
  linker.
  (KP Singh,
  [link](https://lore.kernel.org/bpf/20201117232929.2156341-1-kpsingh@chromium.org/t/#u))

* Switch from memlock rlimit accounting to cgroup-based memory for the kernel

>   memory used by eBPF objects. The rlimit had a number of downsides (please
>   refer to the cover letter linked below for details), and the cgroup-based

Reviewer: nit:

```suggestion
  memory used by eBPF objects. The rlimit had a number of downsides (
  refer to the cover letter linked below for details), and the cgroup-based
```

  offering easier way to retrieve the current amount of memory used. It should
  also reflect the actual memory consumption, while this is not necessarily the
  case with rlimit.
  (Roman Gushchin,
  [link](https://lore.kernel.org/bpf/20201201215900.3569844-1-guro@fb.com/t/#u))

* Allow `bpf_getsockopt()` and `bpf_setsockopt()` helpers from
  `BPF_CGROUP_INET4_BIND` and `BPF_CGROUP_INET6_BIND` attach hooks, so that the
  listener sockets attached to cgroups can pre-populate some options as needed.
  (Stanislav Fomichev,
  [link](https://lore.kernel.org/bpf/20201202172516.3483656-1-sdf@google.com/t/#u))

* Add a new helper `bpf_ktime_get_coarse_ns()` to get a timestamp using the
  `CLOCK_MONOTONIC_COARSE`, which is less accurate but more performant than
  `CLOCK_MONOTONIC` used by `bpf_ktime_get_ns()`.
  (Dmitrii Banshchikov,
  [link](https://lore.kernel.org/bpf/20201117184549.257280-1-me@ubique.spb.ru/t/#u))

This was followed by
[a third pull request](https://lore.kernel.org/bpf/20201214214316.20642-1-daniel@iogearbox.net/),
for the same tree and cycle, and for which you will find the highlights below:

* Expose `bpf_sk_storage_get()` and `bpf_sk_storage_delete()` helpers to

>   iterator programs, for example to initialize or delete selected value from a

Reviewer 4: nit: to, for example, initialize

  (Weqaar Janjua,
  [link](https://lore.kernel.org/bpf/20201207215333.11586-1-weqaar.a.janjua@intel.com/t/#u))

* Update libbpf function `bpf_program__set_attach_target()` so it supports
  finding BTF-based kernel attach targets (such as fentry, fexit, BTF-based
  tracepoints, etc.). Like for other hooks, the attach points can be passed
  through the ELF section name where the eBPF program is placed, with the
  `SEC()` macro.
  (Andrii Nakryiko,
  [link](https://lore.kernel.org/bpf/20201211215825.3646154-1-andrii@kernel.org/t/#u))

* Permit pointers on stack for helper calls in the verifier, if the user has
  sufficient permissions. This addresses an issue where the verifier would
  wrongly reject some programs.
  (Yonghong Song,
  [link](https://lore.kernel.org/bpf/20201210013348.943623-1-yhs@fb.com/t/#u))

* Add new libbpf API function to retrieve an eBPF ring buffer epoll file
  descriptor. This is to help with the migration from perf buffer to eBPF ring
  buffer.
  (Brendan Jackman,
  [link](https://lore.kernel.org/bpf/20201214113812.305274-1-jackmanb@google.com/t/#u))

Kernel 5.10 was released on the 13th of December, bringing a number of eBPF

> novelties. eBPF iterators with preload at boot time, or sleep-able eBPF

Reviewer 3: ```suggestion
novelties. eBPF iterators with preload at boot time, or sleepable eBPF
```

programs, among others. Here I am deferring to [the list on the
Kernelnewbies](https://kernelnewbies.org/Linux_5.10#Tracing.2C_perf_and_BPF)
website for the detailed set of additions.

## Did You Know?: CO-RE

CO-RE (_Compile Once, Run Everywhere_) is a mechanism used with eBPF to ensure
portability of the programs, mainly those intended for tracing. It addresses
the issue that arises when a given structure is modified between two kernel

> versions. Tracing programs may attempt to access to a field from a given

Reviewer 3: ```suggestion
versions. Tracing programs may attempt to access a field from a given
```

structure by reading at a specific offset in that structure. But if the length

> of the structure, or the order of its member fields, for example, is modified

Reviewer 3: I'd drop the `for example` because (1) I can't think of other cases and (2) it becomes a bit heavy in parenthetic clauses (& commas).

Author: For 1), from Andrii's post:
> Different kernel versions will have struct fields shuffled around inside a struct, or even moved into a new inner struct. Fields can be renamed or removed, their types changed, either into some trivially-compatible ones or completely different ones. Structs and other types can get renamed, or they can be conditionally compiled out (depending on kernel configuration), or just plain removed between kernel versions.

Reviewer 3: Ah, right. I forgot about types :+1:

in a different kernel version, then the program will not be compatible.

CO-RE solves this by relying on BTF objects (_BPF Type Format_). A BTF object

> contain debug information about a program. This is in fact a simplified version

Reviewer: ```suggestion
contains debug information about a program. This is in fact a simplified version
```

of the DWARF format, used for example by GDB. BTF objects are loaded into the
kernel, and they often hold information on eBPF bytecode, for example to dump
the C instructions from which the program was compiled. But BTF can also

> describes other objects like the kernel itself. In that case, BTF can provide,

Reviewer 3: ```suggestion
describe other objects like the kernel itself. In that case, BTF can provide,
```

just before an eBPF program is loaded, the relevant information for accessing
the kernel structures. The required adjustments are performed as “ELF
relocation” steps, just before the bytecode is sent to the kernel.

The easiest way to use CO-RE is through libbpf. Note that because it needs the
BTF information for the kernel, CO-RE is only available if the
`CONFIG_DEBUG_INFO_BTF` option has been set when compiling the kernel. Recent
tooling (LLVM, libbpf) is also necessary.

If you want more information on this topic, Andrii Nakryiko explains everything
there is to know about CO-RE
[on his blog](https://nakryiko.com/posts/bpf-portability-and-co-re/).
The specification for BTF can be found in
[the kernel documentation](https://www.kernel.org/doc/html/latest/bpf/btf.html),

> but you might also want to have a look at

Reviewer 3: ```suggestion
but you might also want to have a look at the
```
